### PR TITLE
Rework racket based runtime/compilation

### PR DESCRIPTION
**Pull request by dolio**
This PR includes the changes that get us about as far as we can easily go for our racket evaluation/compilation at the moment. It includes the following aspects.

- I renamed the `runner` file/executable to `ucr`, so that it's more similar to `ucm`. It's not something that people should be calling directly, but `runner` is obviously not a very good name.
- I changed the 'compilation' part of `ucr` to just generate a file. The reason for this is that somehow a math library we are using for the primops causes `create-embedding-executable` to fail. The easiest solution to this problem is to divide the compilation process into two steps: first generate the file using all the unison infrastructure, then compile that file. At first I was thinking about copying the math operations into the primops, but I realized that dividing the programs up is a better solution.
- At the moment, compiling the generated file is just done via `raco exe` using a racket installation. The reason for _this_ is that `create-embedding-executable` does not seem to work in a standalone executable. The API needs to find a copy of the racket executable to modify, and the procedure for that doesn't work in a standalone binary (confirmed with racket devs). Peeking at the code, it doesn't seem like this absolutely _couldn't_ work, but making it work would require someone digging into the racket internals to make it happen. So for the time being, compiling requires a racket install.
- I've included a short transcript that will download `base`, `@unison/internal`, and generate the racket modules alongside the checked-in modules. This is a step necessary in building `ucr` or getting things set up for compilation, so we'll probably use it when we get this stuff building in CI.

Even for compilation, where we still depend on a racket install, we no longer actually need to make people download `@unison/internal` themselves. We can run the library generation as part of a build, and just distribute the generated libraries to people. The runtime should be completely self-contained, though. Once `ucr` is built, `ucm` just needs to shell out to it to execute `run.native`. A racket install is only necessary to _build_ `ucr`.

I think one tweak I have left is a better error message for `compile.native` if `raco` doesn't exist. Beyond that, I just need a little advice on how to get this hooked into CI to get everything distributed properly. I might need to make the native runtime calls a bit more configurable with respect to the location of `ucr` and such, too, but that shouldn't be a big deal.

### CODE REVIEW

**Comment by aryairani** [comment]
> * I renamed the `runner` file/executable to `ucr`, so that it's more similar to `ucm`. It's not something that people should be calling directly, but `runner` is obviously not a very good name.

I'm thinking we could use a more descriptive name and stick it in `libexec` if nobody needs to call it.

> * I changed the 'compilation' part of `ucr` to just generate a file.

What did it do before?

> * I've included a short transcript that will download `base`, `@unison/internal`, and generate the racket modules alongside the checked-in modules. This is a step necessary in building `ucr` or getting things set up for compilation, so we'll probably use it when we get this stuff building in CI.
> 
> Even for compilation, where we still depend on a racket install, we no longer actually need to make people download `@unison/internal` themselves. We can run the library generation as part of a build, and just distribute the generated libraries to people. The runtime should be completely self-contained, though. Once `ucr` is built, `ucm` just needs to shell out to it to execute `run.native`. A racket install is only necessary to _build_ `ucr`.

Can we just do all of this in CI?  Build `ucr` there, and nobody needs to install racket outside of CI?

> I think one tweak I have left is a better error message for `compile.native` if `raco` doesn't exist. Beyond that, I just need a little advice on how to get this hooked into CI to get everything distributed properly. 

We can pair on the CI part if you want.

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 1]
```diff
@@ -2071,21 +2071,21 @@ runScheme file args = do
...
+--   buildRacket genDir statDir main file
+
+-- buildRacket :: String -> String -> String -> String -> Cli ()
+-- buildRacket genDir statDir main file =
+--   let args = ["-l", "raco", "--", "exe", "-o", main, file]
+--       opts = racketOpts genDir statDir args
+--    in void . liftIO $
+--         catch
+--           (True <$ callProcess "racket" opts)
+--           (\(_ :: IOException) -> pure False)
```
Comment (aryairani) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs]: Keep these in for now for reference? Or accidental inclusion?

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 2]
```diff
@@ -2111,9 +2111,9 @@ doRunAsScheme main0 args = case HQ.fromString main0 of
     runScheme fullpath args
   Nothing -> Cli.respond $ BadName main0
 
-doCompileScheme :: String -> HQ.HashQualified Name -> Cli ()
-doCompileScheme out main =
-  generateSchemeFile True out main >>= buildScheme out
+-- doCompileScheme :: String -> HQ.HashQualified Name -> Cli ()
+-- doCompileScheme out main =
+--   generateSchemeFile True out main >>= buildScheme out
```
Comment (aryairani) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs]: Same

**Review by aryairani — APPROVED** [review]

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs (hunk 1 above)
Comment (dolio) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput.hs]: Just haven't deleted them yet.

**Comment by dolio** [comment]
> I'm thinking we could use a more descriptive name and stick it in `libexec` if nobody needs to call it.

Any suggestion on the name? I guess it could be something more like `unison-runtime` or something.

> What did it do before?

It generated a file then threw errors. But the command line gave you the expectation that it would produce an executable.

> Can we just do all of this in CI? Build `ucr` there, and nobody needs to install racket outside of CI?

You can do the equivalent of `run`. Or you can generate a racket source file, but that's useless without racket. If you want to build a standalone executable from some unison code, you need racket installed, because the API for generating executables doesn't work when you build it into a standalone executable.

File: parser-typechecker/src/Unison/Runtime/Interface.hs [hunk 3]
```diff
@@ -696,23 +704,43 @@ nativeEvalInContext _ ctx codes base = do
...
         BS.hPut pin bytes
         UnliftIO.hClose pin
         waitForProcess ph
         pure ()
       callout _ _ _ _ = fail "withCreateProcess didn't provide handles"
-  withCreateProcess (schemeProc ["-o", path]) callout
+      ucrError (_ :: IOException) =
+        die
+          "I had trouble calling the unison runtime exectuable.\n\n\
+          \Please check that the `ucr` executable is properly\
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Runtime/Interface.hs]: Would be great to link to a guide in these errors, as an end user I wouldn't know where to start on ensuring "ucr" is installed and google probably won't help.

Comment (dolio) [reply parser-typechecker/src/Unison/Runtime/Interface.hs]: Yeah, I'd like to make them better. I'm not exactly sure what the install process will be right now, though. Hopefully I can flesh these out after I work out some details with Arya.

**Comment by ChrisPenner** [comment]
Looks great! I'd second the idea of changing `ucr` to something like `unison-runtime`

**Comment by aryairani** [comment]
When would `racoErrMsg` show up currently? Current plan is to never have to call it from within `ucm`, right?  (It doesn't need to be deleted yet, but maybe once CI is working?)

**Comment by dolio** [comment]
`racoErrMsg` shows up when you run `compile.native` and racket isn't installed. You need racket to compile standalone executables right now.

File: parser-typechecker/src/Unison/Runtime/Interface.hs (hunk 3 above)
Comment (aryairani) [reply parser-typechecker/src/Unison/Runtime/Interface.hs]: I'm anticipating that either our installers will set it up correctly for them, or they're in big trouble and need to come ask for help, which will help us flesh out better error messages :-\

**Comment by aryairani** [comment]
Heads up that I'm seeing some intermittent failures in the interpreter tests on macOS.
https://github.com/unisonweb/unison/actions/runs/7937787444/job/21675781417?pr=4675#step:16:173

Any chance that's related to this PR? Otherwise I can hope for green and open a separate ticket.

**Comment by aryairani** [comment]
Note to self: Currently the interpreter tests are being run in the "build JIT binary" stage, which is a little wrong.
But also maybe that's fine / easier because they both rely on the result of `base.md` and it might be a pain to factor that out

**Review by aryairani — CHANGES_REQUESTED** [review]
There's no way to remove an approval (just to make sure it's not merged prematurely); so I'm marking it as requesting changes.

I'm still working on updated drafts of the release scripts, and Dan is changing the behavior of `Nat`.

**Review by aryairani — CHANGES_REQUESTED** [review]
There's no way to remove an approval (just to make sure it's not merged prematurely); so I'm marking it as requesting changes.

I'm still working on updated drafts of the release scripts, and Dan is changing the behavior of `Nat`.

**Comment by dolio** [comment]
Okay, I think the tests in `@aryairani/jit-tests` should pass with this branch.

I also added `compiler.generateSchemeBoot` to `@unison/internal` in the new release `0.0.11` that this depends on. I think you were planning on doing something with that in the transcripts?

**Review by aryairani — APPROVED** [review]
I'm gonna doublecheck CI again in the morning, but I think this is ready to merge even though we may do some additional tweaks later.

**Comment by aryairani** [comment]
> Okay, I think the tests in `@aryairani/jit-tests` should pass with this branch.
> 
> I also added `compiler.generateSchemeBoot` to `@unison/internal` in the new release `0.0.11` that this depends on. I think you were planning on doing something with that in the transcripts?

Thanks for the reminder, got it in 20a0e73.

**Comment by aryairani** [comment]
~The default runtime location is set in `fixNativeRuntimePath`, currently as `$XDG_DATA/unisonlanguage/libexec/unison-runtime`, and can be overridden by `nativeRuntimePathFlag`.~

- [x] It should probably default to a system-wide location though, like `$0/unison-runtime`?

The default runtime location is set in `fixNativeRuntimePath`, as `$0/runtime/unison-runtime[.exe]`, and can be overridden by `nativeRuntimePathFlag`.

**Comment by aryairani** [comment]
> A racket install is only necessary to build `ucr`.

The racket install is still needed to use `compile.native`, right?

**Pull request merged**